File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        pMap = {'{':'}', '[':']', '(':')'}
        stack = []
        for c in s:
            if c not in pMap:
                if not stack:
                    return False
                top = stack.pop()
                if pMap[top] != c:
                    return False
            else:
                stack.append(c)
                
        return  len(stack) == 0

                

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        std::stack<char> stack;
        std::unordered_map<char, char> pairs = {
            {')', '('}, 
            {']', '['}, 
            {'}', '{'}
        };        
        for (char c : s) {
            auto it = pairs.find(c);
            if (it == pairs.end()) {
                stack.push(c);
            }
            else {
                if(stack.empty()) return false;
                if (stack.top() == pairs[c]) stack.pop();
                else break;
            }
        }
        return stack.empty();
    }
};